common/test/java/com/couchbase/lite/ReplicatorMiscTest.java

```diff
@@ -88,23 +91,63 @@ public void testStatusBeforeStart() throws URISyntaxException {
         newReplicator(makeConfig(true, false, false, new URLEndpoint(new URI("wss://foo")))).getStatus();
     }
 
+    @Test
+    public void testGetHeartbeatBeforeSet() throws URISyntaxException {
+        final ReplicatorConfiguration config
+            = new ReplicatorConfiguration(baseTestDb, new URLEndpoint(new URI("wss://foo")));
+        assertEquals(AbstractCBLWebSocket.DEFAULT_HEARTBEAT_SEC, config.getHeartbeat());
+    }
+
+    @Test(expected = IllegalArgumentException.class)
+    public void testSetIllegalHeartbeat() throws URISyntaxException {
+        final ReplicatorConfiguration config
+            = new ReplicatorConfiguration(baseTestDb, new URLEndpoint(new URI("wss://foo")));
+        config.setHeartbeat(-47);
+    }
+
+    @Test(expected = IllegalArgumentException.class)
+    public void testSetZeroHeartbeat() throws URISyntaxException {
+        final ReplicatorConfiguration config
+            = new ReplicatorConfiguration(baseTestDb, new URLEndpoint(new URI("wss://foo")));
+        config.setHeartbeat(-47);
+    }
+
     @Test
     public void testDefaultHeartbeat() throws URISyntaxException {
         final ReplicatorConfiguration config
             = new ReplicatorConfiguration(baseTestDb, new URLEndpoint(new URI("wss://foo")))
             .setReplicatorType(getReplicatorType(true, false))
             .setContinuous(false);
 
-        final Replicator repl = new Replicator(null, config);
+        Replicator repl = new Replicator(null, config);
 
-        final AtomicReference<C4Socket> socketRef = new AtomicReference<>();
-        repl.getSocketFactory().setListener(socketRef::set);
-        try { run(repl); }
-        catch (CouchbaseLiteException ignore) { }
+        final CountDownLatch latch = new CountDownLatch(1);
+        repl.addChangeListener(status -> {
+            if (status.getStatus().getActivityLevel() == AbstractReplicator.ActivityLevel.BUSY) {
+                latch.countDown();
+            }
+        });
 
-        assertEquals(
-            AbstractCBLWebSocket.DEFAULT_HEARTBEAT_SEC * 1000,
-            ((AbstractCBLWebSocket) socketRef.get()).getHttpClient().pingIntervalMillis());
+        final AtomicReference<C4Socket> socketRef = new AtomicReference<>();
+        repl.getSocketFactory().setListener(socket -> socketRef.compareAndSet(null, socket));
+
+        repl.start();
+        try {
+            try { latch.await(STD_TIMEOUT_SECS, TimeUnit.SECONDS); }
+            catch (InterruptedException ignore) { }
+
+            final C4Socket c4socket = socketRef.get();
+            assertNotNull(c4socket);
+            assertTrue(c4socket instanceof AbstractCBLWebSocket);
+            final AbstractCBLWebSocket socket = (AbstractCBLWebSocket) c4socket;
+
+            assertEquals(
+                AbstractCBLWebSocket.DEFAULT_HEARTBEAT_SEC * 1000,
+                socket.getHttpClient().pingIntervalMillis());
+        }
+        finally {
+            repl.stop();
```

Comment:
Is it OK to not wait until the replicator's stopped before returning?

Reply:
After thought, I guessed it's OK, because the tear down will delete or close database and it will wait for the replicator to stop.

Reply:
🤞